### PR TITLE
Exercise2

### PR DESCRIPTION
**Pull request by jobegrabber**

### CODE REVIEW

**Comment by s0557917** [comment]
Hey, the work for the current homework looks good, but if I saw correctly, you also added all the merged files from the homework repository to the PR. It is obviously not wrong, but it makes the list of files longer, making it a bit more tedious to check. It would be better, to either merge it directly with main or create a new branch for it, thus keeping the dedicated feature branch a bit cleaner.
Hope this helps, cheers! 😀

P.S. habe euch zu Reviewers in unser Repo gesetzt, hoffe das ist jetzt in Ordnung.

**Comment by jobegrabber** [comment]
> Hey, the work for the current homework looks good, but if I saw correctly, you also added all the merged files from the homework repository to the PR. It is obviously not wrong, but it makes the list of files longer, making it a bit more tedious to check. It would be better, to either merge it directly with main or create a new branch for it, thus keeping the dedicated feature branch a bit cleaner.
> Hope this helps, cheers! grinning
> 
> P.S. habe euch zu Reviewers in unser Repo gesetzt, hoffe das ist jetzt in Ordnung.

Yes, thanks, we should merge the new exercise into `main` and rebase on that!
**EDIT:** Done.

We're peer-reviewing with team [lichtow](https://github.com/Systems-Development-and-Frameworks/lichtow) and are fairly constrained in terms of time available. Not sure if you picked the wrong group you wanted to request a review from.
If you didn't find a group to get a review from I think you should approach @medizinmensch and @roschaefer first.

**Comment by s0557917** [comment]
> 
> 
> > Hey, the work for the current homework looks good, but if I saw correctly, you also added all the merged files from the homework repository to the PR. It is obviously not wrong, but it makes the list of files longer, making it a bit more tedious to check. It would be better, to either merge it directly with main or create a new branch for it, thus keeping the dedicated feature branch a bit cleaner.
> > Hope this helps, cheers! grinning
> > P.S. habe euch zu Reviewers in unser Repo gesetzt, hoffe das ist jetzt in Ordnung.
> 
> Yes, thanks, we should merge the new exercise into `main` and rebase on that!
> **EDIT:** Done.
> 
> We're peer-reviewing with team [lichtow](https://github.com/Systems-Development-and-Frameworks/lichtow) and are fairly constrained in terms of time available. Not sure if you picked the wrong group you wanted to request a review from.
> If you didn't find a group to get a review from I think you should approach @medizinmensch and @roschaefer first.

Allright, thanks! I'll ask around then.

File: .github/workflows/node.js.yml [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+jobs:
+  build-test-lint:
+
+    runs-on: ubuntu-latest
+
+    strategy:
+      matrix:
+        node-version: [10.x, 12.x, 14.x]
+
```
Comment (noeljns) [inline .github/workflows/node.js.yml:24]: Any reason why you choose three node versions instead of only one?

File: .github/workflows/node.js.yml [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+    working-directory: ./webapp/dojo_news
+
+jobs:
+  build-test-lint:
+
+    runs-on: ubuntu-latest
+
+    strategy:
+      matrix:
+        node-version: [10.x, 12.x, 14.x]
```
Comment (noeljns) [inline .github/workflows/node.js.yml:23]: Any reason why you choose three versions instead of specifying one?

File: README.md [hunk 3]
```diff
@@ -1,8 +1,39 @@
-# Systems Development and Frameworks - 2020/21
+# Systems Development and Frameworks - 2020/21 [![Build status](https://github.com/Systems-Development-and-Frameworks/dojo/workflows/Build%2C%20Test%2C%20Lint/badge.svg?branch=exercise2)](https://github.com/Systems-Development-and-Frameworks/dojo/actions?query=workflow%3A%22Build%2C+Test%2C+Lint%22+branch%3Aexercise2)
```
Comment (powlaa) [inline README.md:1]: Really cool that you added the status in the Readme!  👍

File: webapp/dojo_news/src/App.spec.js [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+        expect(wrapperWithEmptyList.find("#reverse-order-button").element.hasAttribute("disabled")).toBeTruthy();
+        expect(wrapperWithEmptyList.find("#reverse-order-button").element.getAttribute("disabled")).toEqual("disabled");
+
+        wrapperWithEmptyList.vm.createNewsListItem("Test");
+        await Vue.nextTick();
+
+        expect(wrapperWithEmptyList.find("#reverse-order-button").element.hasAttribute("disabled")).not.toBeTruthy();
+        expect(wrapperWithEmptyList.find("#reverse-order-button").element.getAttribute("disabled")).not.toEqual("disabled");
+    });
+});
```
Comment (noeljns) [inline webapp/dojo_news/src/App.spec.js]: Nice that you cover so many different test cases.

File: webapp/dojo_news/src/components/NewsListItem.stories.js [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+PositiveScore.args = {
+    title: "Positive Score NewsListItem",
+    votes: 4,
+};
+
+export const NegativeScore = Template.bind({});
+NegativeScore.args = {
+    title: "Negative Score NewsListItem",
+    votes: -2,
+};
```
Comment (noeljns) [inline webapp/dojo_news/src/components/NewsListItem.stories.js:42]: Great that you have more const stories than just a default!

**Review by noeljns — APPROVED** [review]

File: webapp/dojo_news/src/App.spec.js [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+
+        wrapper.vm.createNewsListItem("Test");
+        wrapper.vm.createNewsListItem("Test2");
+
+        await Vue.nextTick();
+
+        expect(wrapper.find("#newslist").html()).not.toBe('<div id="newslist"></div>');
+        expect(wrapper.findAll(".newslistitem").length).toBe(2);
+    });
+});
```
Comment (powlaa) [inline webapp/dojo_news/src/App.spec.js]: There are 2 things that are being tested here, adding items and removing items. I think it would be better to separate those into two test cases so that it will be more clear where a test fails.

File: webapp/dojo_news/src/App.vue [hunk 7]
```diff
@@ -23,29 +31,54 @@ import NewsListItemInput from "./components/NewsListItemInput";
...
   components: {
     NewsListItem,
-    NewsListItemInput
+    NewsListItemInput,
+  },
+  props: {
+    listEmptyMessage: {
+      type: String,
+      default: "The list is empty :(",
+    },
```
Comment (powlaa) [inline webapp/dojo_news/src/App.vue:40]: I like that you added listEmptyMessage as a prop as well 👍

**Review by powlaa — APPROVED** [review]
Overall I really like how you implemented all the tasks from exercise 2  👍 

I'm not sure why but there seem to be a lot of files in 'archive/' that have been deleted and show up in the Files changed as well.  (It was a  bit annoying to click through them all when reviewing)

File: README.md [hunk 8]
```diff
@@ -1,8 +1,39 @@
...
+### Compiles and minifies for production
+```
+$ (webapp) npm run build
+```
+
+### Lints and fixes files
+```
+$ (webapp) npm run lint
+```
+
```
Comment (roschaefer) [inline README.md:35]: Absolutely :heart: that.

File: webapp/dojo_news/src/App.spec.js [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+                initialNewsListItems: [
+                    {id: 0, title: "macOS", votes: 0},
+                    {id: 1, title: "Linux", votes: 0},
+                    {id: 2, title: "Windows", votes: 0}
+                ]
+            }
+        });
+        expect(wrapper.find("#newslist").html()).not.toBe('<div id="newslist"></div>');
+
+        expect(wrapper.findAll(".newslistitem").length).toBe(3);
```
Comment (roschaefer) [inline webapp/dojo_news/src/App.spec.js]:
```suggestion
        expect(wrapper.findAll(".newslistitem")).toHaveLength(3);
```

File: webapp/dojo_news/src/App.spec.js [hunk 10]
```diff
@@ -0,0 +1,221 @@
...
+        const wrapper = shallowMount(App, {
+            propsData: {
+                initialNewsListItems: [
+                    {id: 0, title: "macOS", votes: 0},
+                    {id: 1, title: "Linux", votes: 0},
+                    {id: 2, title: "Windows", votes: 0}
+                ]
+            }
+        });
+        expect(wrapper.find("#newslist").html()).not.toBe('<div id="newslist"></div>');
```
Comment (roschaefer) [inline webapp/dojo_news/src/App.spec.js]:
```suggestion
```

Testing nothingness is really error-prone and can lead to false negatives. If you have to, use
```
Arrange - Assert - Act - Assert
```
to avoid silently passing tests.

File: webapp/dojo_news/src/App.spec.js [hunk 11]
```diff
@@ -0,0 +1,221 @@
...
+import Vue from "vue";
+
+describe("News list", () => {
+    it('does not render any items on an initially empty list of NewsItems', () => {
+        const wrapper = shallowMount(App, {
+            propsData: {
+                initialNewsListItems: []
+            }
+        });
+        expect(wrapper.find("#newslist").html()).toBe('<div id="newslist"></div>');
```
Comment (roschaefer) [inline webapp/dojo_news/src/App.spec.js]: You could use [toBeEmpty](https://github.com/testing-library/jest-dom#tobeempty) to make this a little more telling.

File: webapp/dojo_news/src/App.spec.js (hunk 9 above)
Comment (roschaefer) [reply webapp/dojo_news/src/App.spec.js]: That's a better assertion than the one above :point_up:

File: webapp/dojo_news/src/App.spec.js [hunk 12]
```diff
@@ -0,0 +1,221 @@
...
+        expect(wrapper.findAll(".newslistitem").length).toBe(0);
```
Comment (roschaefer) [inline webapp/dojo_news/src/App.spec.js]: I would try to do one assertion per test case.

I would prefer
```js
        expect(wrapper.findAll(".newslistitem")).toHaveLength(0);
```
over
```js
        expect(wrapper.find("#newslist").html()).toBe('<div id="newslist"></div>');
```

Or make two test cases!

File: webapp/dojo_news/src/App.spec.js [hunk 13]
```diff
@@ -0,0 +1,221 @@
...
+    });
+    it('renders items correctly dependent on mutations of the list of NewsItems', async () => {
+        const wrapper = shallowMount(App, {
+            propsData: {
+                initialNewsListItems: [
+                    {id: 0, title: "macOS", votes: 0},
+                ]
+            }
+        });
+        expect(wrapper.find("#newslist").html()).not.toBe('<div id="newslist"></div>');
```
Comment (roschaefer) [inline webapp/dojo_news/src/App.spec.js]:
```suggestion
```

File: webapp/dojo_news/src/App.spec.js [hunk 14]
```diff
@@ -0,0 +1,221 @@
...
+        expect(wrapper.findAll(".newslistitem").length).toBe(1);
+
+        wrapper.vm.removeNewsListItem(0);
```
Comment (roschaefer) [inline webapp/dojo_news/src/App.spec.js:40]: No! Don't mess around with internal properties/methods.

Better:
```js
await wrapper.find('button').trigger('click')
```

You would need to use `mount` instead of `shallowMount`

For the `shallowMount` version:

```js
wrapper.find(SomeComponent).vm.$emit('someEvent')
```

Unfortunately, `wrapper.trigger(event)` only works for native DOM events. https://github.com/vuejs/vue-test-utils/issues/145

File: webapp/dojo_news/src/App.spec.js [hunk 15]
```diff
@@ -0,0 +1,221 @@
...
+
+        await Vue.nextTick();
```
Comment (roschaefer) [inline webapp/dojo_news/src/App.spec.js:42]: This is obsolete when you `await` the `trigger(event)`.

File: webapp/dojo_news/src/App.spec.js [hunk 16]
```diff
@@ -0,0 +1,221 @@
...
+
+        expect(wrapper.find("#newslist").html()).toBe('<div id="newslist"></div>');
+        expect(wrapper.findAll(".newslistitem").length).toBe(0);
+
+        wrapper.vm.createNewsListItem("Test");
```
Comment (roschaefer) [inline webapp/dojo_news/src/App.spec.js:57]: Use `mount` and then properly fill in some input and click a button:

https://vue-test-utils.vuejs.org/api/wrapper/#setvalue
https://vue-test-utils.vuejs.org/api/wrapper/#trigger

File: webapp/dojo_news/src/App.spec.js (hunk 6 above)
Comment (roschaefer) [reply webapp/dojo_news/src/App.spec.js]: Yes, and also DRY your test case descriptions a bit:
```
  News list
    ✓ does not render any items on an initially empty list of NewsItems (29 ms)
    ✓ renders items on an initially filled list of NewsItems (8 ms)
    ✓ renders items correctly dependent on mutations of the list of NewsItems (12 ms)
  List Empty Message
    ✓ displays a message on an initially empty list of NewsItems (3 ms)
    ✓ displays no message on an initially non-empty list of NewsItems (4 ms)
    ✓ adds and removes the message on list mutations (6 ms)
  Reverse Order Button
    ✓ list order is reversed upon Reverse-Button click from initial state _descending_ (8 ms)
    ✓ list order is reversed upon Reverse-Button click from initial state _ascending_ (6 ms)
    ✓ displays the Reverse-Order-Button on an initially non empty list of NewsItems (5 ms)
    ✓ does not display the Reverse-Order-Button on an initially empty list of NewsItems (4 ms)
    ✓ Reverse-Button is hidden upon empty list (6 ms)
    ✓ Reverse-Button shows up again upon non-empty list (6 ms)
```

What about something like:
```
  News list
    given no items
      renders empty state
```
```

File: webapp/dojo_news/src/App.spec.js [hunk 17]
```diff
@@ -0,0 +1,221 @@
...
+    const listEmptyMessage = "Oh noes, the list is empty!";
+
+    it('displays a message on an initially empty list of NewsItems', () => {
+        const wrapperWithEmptyList = shallowMount(App, {
+            propsData: {
+                listEmptyMessage: listEmptyMessage,
+                initialNewsListItems: []
+            }
+        });
+        expect(wrapperWithEmptyList.text()).toContain(listEmptyMessage);
```
Comment (roschaefer) [inline webapp/dojo_news/src/App.spec.js]:
```suggestion
        expect(wrapperWithEmptyList.text()).toContain("Oh noes, the list is empty!");
```

I prefer hard-coded assertion values.

File: webapp/dojo_news/src/App.spec.js [hunk 18]
```diff
@@ -0,0 +1,221 @@
...
+    });
+
+    it('displays no message on an initially non-empty list of NewsItems', () => {
+        const wrapper = shallowMount(App, {
```
Comment (roschaefer) [inline webapp/dojo_news/src/App.spec.js:77]: I like to use a setup method:
```js
const setupWrapper = (opts) => {
  const defaults = { /* ... */ }
  shallowMount(App, {...defaults, ...opts })
}
```

In you describe block:
```js
let propsData
describe('given no `items`', () => {
  beforeEach(() => {
    propsData = []
  })
})
```

In your test case:
```js
it('renders empty state', () => {
  wrapper = setupWrapper({ propsData })
})
```

File: webapp/dojo_news/src/App.spec.js [hunk 19]
```diff
@@ -0,0 +1,221 @@
...
+
+        expect(wrapper.text()).not.toContain(listEmptyMessage);
+
+        wrapper.vm.removeNewsListItem(0);
+
+        await Vue.nextTick();
+
+        expect(wrapper.html()).toContain(listEmptyMessage);
+    });
+});
```
Comment (roschaefer) [inline webapp/dojo_news/src/App.spec.js]: If your test case description already has `do X and Y` then probably you should split the test case.

File: webapp/dojo_news/src/App.spec.js [hunk 20]
```diff
@@ -0,0 +1,221 @@
...
+                    {id: 2, title: "centOS", votes: 3},
+                    {id: 3, title: "Windows", votes: 0}
+                ],
+                initialDescendingOrder: true
+            }
+        });
+
+        let expectedIdList = [1, 0, 2, 3];
+        // or check HTML
+        let actualIdList = wrapper.vm.sortedNewsListItems.map(x => x.id);
```
Comment (roschaefer) [inline webapp/dojo_news/src/App.spec.js]: Here you're testing against the internal property`sortedNewsListItems` and not what's being rendered. This puts you under a high risk of false-negatives.

File: webapp/dojo_news/src/App.spec.js [hunk 21]
```diff
@@ -0,0 +1,221 @@
...
+
+    it('does not display the Reverse-Order-Button on an initially empty list of NewsItems', () => {
+        const wrapperWithEmptyList = shallowMount(App, {
+            propsData: {
+                initialNewsListItems: []
+            }
+        });
+
+        expect(wrapperWithEmptyList.find("#reverse-order-button").element.hasAttribute("disabled")).toBeTruthy();
+        expect(wrapperWithEmptyList.find("#reverse-order-button").element.getAttribute("disabled")).toEqual("disabled");
```
Comment (roschaefer) [inline webapp/dojo_news/src/App.spec.js]:
```suggestion
        expect(wrapperWithEmptyList.find("#reverse-order-button").attributes("disabled")).toEqual("disabled");
```
See: https://vue-test-utils.vuejs.org/api/wrapper/#attributes

File: webapp/dojo_news/src/App.vue [hunk 22]
```diff
@@ -1,18 +1,26 @@
...
+            :title="item.title"
+            :votes="item.votes"
+            @upvote="item.votes++"
+            @downvote="item.votes--"
+            @remove="removeNewsListItem(item.id)"
+        />
+      </div>
     </div>
     <NewsListItemInput @create="createNewsListItem($event);"/>
+    <button type="button" @click="toggleOrder()" :disabled="!newsListItems.length" id="reverse-order-button">Reverse Order
```
Comment (roschaefer) [inline webapp/dojo_news/src/App.vue]:
```suggestion
    <button type="button" @click="toggleOrder" :disabled="!newsListItems.length" id="reverse-order-button">Reverse Order
```

File: webapp/dojo_news/src/App.vue [hunk 23]
```diff
@@ -23,29 +31,54 @@ import NewsListItemInput from "./components/NewsListItemInput";
...
+      this.descendingOrder = !this.descendingOrder;
     }
   },
   computed: {
     sortedNewsListItems() {
-      return [...this.newsListItems].sort((a, b) => b.votes - a.votes);
+      const compareFn = this.descendingOrder
+          ? (a, b) => b.votes - a.votes
+          : (a, b) => a.votes - b.votes
+      return [...this.newsListItems].sort(compareFn);
```
Comment (roschaefer) [inline webapp/dojo_news/src/App.vue:81]: :+1:

**Review by roschaefer — CHANGES_REQUESTED** [review]
perfect

Your PR is close to perfect. Only your test cases and implementation could be improved, see my suggestions below.

You got all 9 :star: plus one :star: for your review here https://media.giphy.com/media/l2Sq5GffrCyUMEXjW/giphy.gif

That's 10/10 :star: well done!

File: webapp/dojo_news/src/App.spec.js (hunk 9 above)
Comment (donatbrzoska) [reply webapp/dojo_news/src/App.spec.js]: Done, thanks!

File: webapp/dojo_news/src/App.spec.js (hunk 10 above)
Comment (donatbrzoska) [reply webapp/dojo_news/src/App.spec.js]: This got fixed in the meantime by removing unnecessary double checks.

File: webapp/dojo_news/src/App.spec.js (hunk 11 above)
Comment (donatbrzoska) [reply webapp/dojo_news/src/App.spec.js]: Done, but with [toBeEmptyDOMElement](https://github.com/testing-library/jest-dom#tobeemptydomelement), thanks!

File: webapp/dojo_news/src/App.spec.js (hunk 12 above)
Comment (donatbrzoska) [reply webapp/dojo_news/src/App.spec.js]: I see your point, in this case one check is totally enough. Fixed it.

File: webapp/dojo_news/src/App.spec.js (hunk 13 above)
Comment (donatbrzoska) [reply webapp/dojo_news/src/App.spec.js]: Done!

File: webapp/dojo_news/src/App.spec.js (hunk 6 above)
Comment (donatbrzoska) [reply webapp/dojo_news/src/App.spec.js]: While I agree with you on writing two seperate tests for two functionalities (got that fixed in the meantime), I think the structure of the tests is more of a personal preference thing. We structured our tests by feature, so we add the state to each of the test case titles. You would structure it by state, but then we would have to add the feature title to it. Or is there any other reason for our version to be bad practice?

File: webapp/dojo_news/src/App.spec.js (hunk 19 above)
Comment (donatbrzoska) [reply webapp/dojo_news/src/App.spec.js]: Fixed!

File: webapp/dojo_news/src/App.spec.js (hunk 17 above)
Comment (donatbrzoska) [reply webapp/dojo_news/src/App.spec.js]: Is that a request for changes or an annotation? :D And wouldn't this also require all the tests to be adjusted, when the list empty message gets changed?

File: webapp/dojo_news/src/App.spec.js (hunk 6 above)
Comment (roschaefer) [reply webapp/dojo_news/src/App.spec.js]: Just make sure your test case descriptions are documenting the behaviour of your code well. It's always a little tricky because it depends on the perception of humans.

I draw lot of my knowledge from sites like https://www.betterspecs.org/

Screenshot - 2020-11-15T140650 199

But yes, I like your idea of structuring your tests by feature.
```
  News list
    emty state
      is displayed when list items are empty
```

File: webapp/dojo_news/src/App.spec.js (hunk 17 above)
Comment (roschaefer) [reply webapp/dojo_news/src/App.spec.js]: Yes, that' correct, you have to touch your tests when you change the empty message.

The reason is a trade-off: I ran into false-negatives in the past when I used assertion values based on a variable. An example is e..g empty string is always contained in any text. If, for some reason, you made a mistake in your test setup you may run into a silently passing (but false) test cases. Also, you lower the complexity of your test case, making it easier to read.

Failing tests because of a changed empty message would be a false-positive in this definition. Yet, false-negatives are a lot *worse* than false-positives, that's why I suggest to use hard-coded assertion values where possible.

File: webapp/dojo_news/src/App.spec.js (hunk 6 above)
Comment (donatbrzoska) [reply webapp/dojo_news/src/App.spec.js]: I definitely see your point, it's a dream to be able to read code and/or test results like that. I renamed all the test cases for a better reading experience of the `npm run test:unit` output.

But I still think that adding another description layer would be too verbose in this case. Because for example, you would have the description "given no items" for every feature, just for one test case. And if I were to do it like that, I also had to add the descriptions "given items", "upon empty news list" and "upon non-empty news list". Everywhere, just for one test case, which in my eyes ist way too verbose.

File: webapp/dojo_news/src/App.vue (hunk 22 above)
Comment (donatbrzoska) [reply webapp/dojo_news/src/App.vue]: Fixed, thanks!

File: webapp/dojo_news/src/App.spec.js (hunk 21 above)
Comment (donatbrzoska) [reply webapp/dojo_news/src/App.spec.js]: Fixed, thanks!

File: webapp/dojo_news/src/App.spec.js (hunk 17 above)
Comment (donatbrzoska) [reply webapp/dojo_news/src/App.spec.js]: "An example is e..g empty string is always contained in any text." - eye opening, thanks! Fixed it. However you probably mean "false-positives" when you write "false-negatives" and vice versa. If you were to check for an empty string by mistake and the test would pass because of that, the result is positive, but that is false. So you have a false-positive, which is indeed a lot worse than a false-negative.

File: webapp/dojo_news/src/App.spec.js (hunk 18 above)
Comment (donatbrzoska) [reply webapp/dojo_news/src/App.spec.js:77]: This way, you would add abstraction to the tests again. Isn't this kind of unwanted? Or better: Where do we want abstraction, where do we not?

File: webapp/dojo_news/src/App.spec.js (hunk 17 above)
Comment (roschaefer) [reply webapp/dojo_news/src/App.spec.js]: Depends on your definition: I use the analogy of a [Corona|HIV]-Test: It's negative when the test says, you are not infected. So a software test is negative, when it's passing. When you have a test case passing when it should be failing, this would be a false-negative according to this definition.

Comment (donatbrzoska) [reply webapp/dojo_news/src/App.spec.js]: Right.

File: webapp/dojo_news/src/App.spec.js (hunk 18 above)
Comment (roschaefer) [reply webapp/dojo_news/src/App.spec.js:77]: It certainly adds abstraction. There a folks out there who prefer an all-flat test hierarchy without any `describe`, `beforeEach` and friends. All code for test setup would have to be copied into every `test('... ', () => {})` case. This adds a lot of redundancy but has the benefit that the reader has to look into the implementation of the test case only to be able to understand the test.

I try to prevent false-negatives and I found an all flat-hierarchy can *also be error-prone*.

When I implement a feature, I do ["red, green, refactor"](https://www.codecademy.com/articles/tdd-red-green-refactor). In addition to the red test, I add one green test case which shares some test setup of the red test. This green test case helps me to know that the test setup is correct. If I had to maintain multiple test setups because I have an all-flat hierarchy, then I can make mistakes when I copy over the code.

A shared test setup for all test cases within a `describe` block helps to avoid this problem. I can always rely on 1 failing / 1 passing case when I change the setup.


Following this 1 green / 1 red pattern, I usually build up a tree-like hierarchy:
```
feature to be tested
  behaves like X
  given a condition A
    behaves like Y
    given condition B
      behaves like Z
    but given condition C
      behaves like X again
```

So I hope this explains why I'm in favour of hard-coded assertion values (low complexity) but also suggest something like a shared setup method (high complexity). In each `beforeEach` inside a `describe('condition A', () => {})` you would only implement changes for condition A.

File: webapp/dojo_news/src/App.spec.js (hunk 20 above)
Comment (donatbrzoska) [reply webapp/dojo_news/src/App.spec.js]: I tried to fix that but it's a pretty ugly solution in my opinion. I am testing against the `wrapper.find("newslist").element.innerHTML`. It feels very wrong because I obtained the "correct" string by printing this exact thing and looking through it for possible errors. And this is still not the final string that actually gets rendered by the browser ... Do you have a better way of doing this in mind?

EDIT: We found another way, commit coming soon!

Comment (roschaefer) [reply webapp/dojo_news/src/App.spec.js]:
```js
expect(wrapper.findAllComponents(NewsItem).wrappers.map(w => w.text()))
  .toEqual(['Title a (2)', 'Title b (1)'])
```

File: .github/workflows/node.js.yml (hunk 2 above)
Comment (jobegrabber) [reply .github/workflows/node.js.yml:23]: These are the [three current LTS versions](https://nodejs.org/en/about/releases/).

Comment (roschaefer) [reply .github/workflows/node.js.yml:23]: On the other PRs I said one version would be enough as you have control over the node version: Either you on your laptop or your build server will transpile your VueJS code.

But no biggie.
